magento/magento2
# Fix bundle upsell missing limit bug

### Description (*)

An observer Magento\Bundle\Observer\AppendUpsellProductsObserver does not honour upsell limit configurations. That causes an issue with an infinite number of products added to the upsell collection and potentially leads to performance issues and bad user experience.
This fix updates "catalog_product_upsell" event dispatching with the config value set in the XML (app/code/Magento/Bundle/view/frontend/layout/catalog_product_view_type_simple.xml). It also includes some improvements to Unit Tests.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#31840

### Manual testing scenarios (*)

1. Create a simple product, to be used as a child in several bundle products
2. Create several bundle products and include a simple product created in the first step into all bundles. The number of bundles should be greater than four as the default number of products configured to be displayed in the upsell block is four.
3. Flush Magento cache.
4. Visit the simple product details page. 
    * Actual result: all bundle products are displayed in the upsell block
    * Expected result:: only four bundle products should be displayed in the upsell block.


### Contribution checklist (*)
